pyteal/ast/abi/tuple.py

```diff
@@ -117,6 +117,99 @@ def _encode_tuple(values: Sequence[BaseType]) -> Expr:
     return Concat(*toConcat)
 
 
+def _index_tuple_bytes(
+    value_types: Sequence[TypeSpec], encoded: Expr, index: int
+) -> Expr:
+    if not (0 <= index < len(value_types)):
+        raise ValueError("Index outside of range")
+
+    offset = 0
+    ignoreNext = 0
+    lastBoolStart = 0
+    lastBoolLength = 0
+    for i, typeBefore in enumerate(value_types[:index]):
+        if ignoreNext > 0:
+            ignoreNext -= 1
+            continue
+
+        if typeBefore == BoolTypeSpec():
+            lastBoolStart = offset
+            lastBoolLength = _consecutive_bool_type_spec_num(value_types, i)
+            offset += _bool_sequence_length(lastBoolLength)
+            ignoreNext = lastBoolLength - 1
+            continue
+
+        if typeBefore.is_dynamic():
+            offset += 2
+            continue
+
+        offset += typeBefore.byte_length_static()
+
+    valueType = value_types[index]
+
+    if type(valueType) is Bool:
+        if ignoreNext > 0:
+            # value is in the middle of a bool sequence
+            bitOffsetInBoolSeq = lastBoolLength - ignoreNext
+            bitOffsetInEncoded = lastBoolStart * NUM_BITS_IN_BYTE + bitOffsetInBoolSeq
+        else:
+            # value is the beginning of a bool sequence (or a single bool)
+            bitOffsetInEncoded = offset * NUM_BITS_IN_BYTE
+        return GetBit(encoded, Int(bitOffsetInEncoded))
```

Comment:
bug here but also where(?) were checking return type

Would it help to have Expr[bytes|uint64] at least for mypy typechecking?

Reply:
I asked Chat GPT to rewrite the function. It came up with the following, which still passes our unit tests. Does this one also have the bug?

```python

def _index_tuple_bytes(value_types, encoded, index):
    if not (0 <= index < len(value_types)):
        raise ValueError("Index outside of range")

    offset = 0
    for i, value_type in enumerate(value_types[:index]):
        if value_type == BoolTypeSpec():
            bool_length = _consecutive_bool_type_spec_num(value_types, i)
            offset += _bool_sequence_length(bool_length)
        elif value_type.is_dynamic():
            offset += 2
        else:
            offset += value_type.byte_length_static()

    value_type = value_types[index]

    if value_type == BoolTypeSpec():
        bit_offset = offset * NUM_BITS_IN_BYTE
        return GetBit(encoded, Int(bit_offset))
    elif value_type.is_dynamic():
        has_next_dynamic_value = False
        next_dynamic_value_offset = offset + 2
        for i, type_after in enumerate(value_types[index + 1 :], start=index + 1):
            if type_after.is_dynamic():
                has_next_dynamic_value = True
                break
            if type_after == BoolTypeSpec():
                bool_length = _consecutive_bool_type_spec_num(value_types, i)
                next_dynamic_value_offset += _bool_sequence_length(bool_length)
            else:
                next_dynamic_value_offset += type_after.byte_length_static()

        start_index = ExtractUint16(encoded, Int(offset))
        if not has_next_dynamic_value:
            return substring_for_decoding(encoded, start_index=start_index)
        end_index = ExtractUint16(encoded, Int(next_dynamic_value_offset))
        return substring_for_decoding(
            encoded, start_index=start_index, end_index=end_index
        )
    else:
        start_index = Int(offset)
        length = Int(value_type.byte_length_static())
        if index + 1 == len(value_types):
            if offset == 0:
                return encoded
            return substring_for_decoding(encoded, start_index=start_index)
        if offset == 0:
            return substring_for_decoding(encoded, length=length)
        return substring_for_decoding(encoded, start_index=start_index, length=length)
```

Reply:
only a bug in that im not checking the return type at least in the `encode` function is getting `Bytes`

It does return ```        return GetBit(encoded, Int(bit_offset))
``` but im not too mad since it couldnt know GetBit would put a uint64 on the stack

Reply:
Right. I see now what you're saying. Yeah, maybe generics could be helpful here. Given the fact that such a type check would get ignored at runtime, I still see us needing our `type_of()` assertions.